### ISSUE TITLE
[C++] the require input Shape of Convolution change while bind

### ISSUE THREAD

**Issue by kice** [dmlc/mxnet#5952]
## Environment info
Operating System: Win10

Compiler: MSVC 15 (CUDAv8 + MSVC14)

Package used (Python/R/Scala/Julia): C++

MXNet commit hash: commit b5a71d5

But I get same error message for the nightly build (20170423_mxnet_x64_vc14_gpu.7z).

## Error Message:
Console output

[23:02:47] D:\mxnet\dmlc-core\include\dmlc/logging.h:300: [23:02:47] D:\mxnet\src\ndarray\ndarray.cc:239: Check failed: from.shape() == to->shape() operands shape mismatchfrom.shape = (64,) to.shape=(64,1,41,41)

And there is a c++ exception will thrown with error message:
Error in operator conv_conv01: Shape inconsistent, Provided=(64,1,3,3), inferred shape=(64,64,3,3)

```

#include "fmt\format.h" // https://github.com/fmtlib/fmt

inline Symbol Convolution(const std::string& symbol_name,
                          Symbol data,
                          Symbol weight,
                          Symbol bias,
                          Shape kernel,
                          uint32_t num_filter,
                          Shape stride = Shape(),
                          Shape dilate = Shape(),
                          Shape pad = Shape(),
                          uint32_t num_group = 1,
                          uint64_t workspace = 1024,
                          bool no_bias = false,
                          ConvolutionCudnnTune cudnn_tune = ConvolutionCudnnTune::None,
                          bool cudnn_off = false,
                          ConvolutionLayout layout = ConvolutionLayout::None) {
  static const char *ConvolutionCudnnTuneValues[] = {
    "None",
    "fastest",
    "limited_workspace",
    "off"
  };
  static const char *ConvolutionLayoutValues[] = {
    "None",
    "NCDHW",
    "NCHW",
    "NCW",
    "NDHWC",
    "NHWC"
  };
  return Operator("Convolution")
           .SetParam("kernel", kernel)
           .SetParam("num_filter", num_filter)
           .SetParam("stride", stride)
           .SetParam("dilate", dilate)
           .SetParam("pad", pad)
           .SetParam("num_group", num_group)
           .SetParam("workspace", workspace)
           .SetParam("no_bias", no_bias)
           .SetParam("cudnn_tune", ConvolutionCudnnTuneValues[int(cudnn_tune)])
           .SetParam("cudnn_off", cudnn_off)
           .SetParam("layout", ConvolutionLayoutValues[int(layout)])
           .SetInput("data", data)
           .SetInput("weight", weight)
           .SetInput("bias", bias)
           .CreateSymbol(symbol_name);
}

inline Symbol ConvFactory(Symbol data, int num_filter,
						  Shape kernel,
						  Shape stride = Shape(1, 1),
						  Shape pad = Shape(0, 0),
						  const std::string & name = "")
{
	Symbol conv_w("conv_" + name + "_w"), conv_b("conv_" + name + "_b");

	Symbol conv = Convolution("conv_" + name, data,
							  conv_w, conv_b, kernel,
							  num_filter, stride, Shape(1, 1), pad);
	return Activation("relu_" + name, conv, "relu");
}

// model
Symbol data = Symbol::Variable("data");
Symbol data_label = Symbol::Variable("data_label");
auto conv1 = ConvFactory(data, 64, Shape(3, 3), Shape(1, 1), Shape(1, 1), "conv01");

std::vector<Symbol> conv;
conv.push_back(conv1);
for (int i = 1; i <= 8; ++i) {
	std::string layerID = fmt::sprintf("conv%02d", i);
	Symbol convx = ConvFactory(conv.back(), 64, Shape(3, 3), Shape(1, 1), Shape(1, 1), layerID);
	conv.push_back(convx);
}
...

// training
auto model = Model();
std::map<std::string, NDArray> args_map;
std::map<std::string, NDArray> aux_map;

args_map["data"] = NDArray(Shape(64, 1, 41, 41), Context::cpu());
args_map["data_label"] = NDArray(Shape(61, 1, 41, 41), Context::cpu());
model.InferArgsMap(ctx, &args_map, args_map); // this line is fine

...

auto train_iter = MXDataIter("ImageRecordIter")
	.SetParam("path_imglist", "./list_folder/train.lst")
	.SetParam("path_imgrec", "./list_folder/train.rec")
	.SetParam("data_shape", Shape(1, 41, 41))
	.SetParam("batch_size", 64)
	.CreateDataIter();

auto val_iter = MXDataIter("ImageRecordIter")
	.SetParam("path_imglist", "./list_folder/val.lst")
	.SetParam("path_imgrec", "./list_folder/val.rec")
	.SetParam("data_shape", Shape(1, 41, 41))
	.SetParam("batch_size", 64)
	.CreateDataIter();

...

	train_iter.Reset();
	while (train_iter.Next()) {
		auto data_batch = train_iter.GetDataBatch();
		auto label = data_batch.label.GetShape();
		data_batch.data.CopyTo(&args_map["data"]);
		data_batch.label.CopyTo(&args_map["data_label"]); // Check failed: from.shape() == to->shape() operands shape mismatchfrom.shape = (64,) to.shape=(64,1,41,41)
		NDArray::WaitAll();

		auto *exec = model.SimpleBind(ctx, args_map); // Error in operator conv_conv01: Shape inconsistent, Provided=(64,1,3,3), inferred shape=(64,64,3,3)
		exec->Forward(true);
		exec->Backward();
		exec->UpdateAll(opt, learning_rate, weight_decay);

		delete exec;
	}

...

```

Most of the code is from examples inside cpp-package.
https://github.com/dmlc/mxnet/blob/master/cpp-package/example/googlenet.cpp and more

I have on idea why it changes the input shape of the Convolution input depth form 1 -> 64.

And I change somethings:
First: I deleted the line:
```
model.InferArgsMap(ctx, &args_map, args_map);
```
I can get the InferExecutorArrays return successful, but the new Executor op has throw another error: 0xC0000005
https://github.com/dmlc/mxnet/blob/master/src/executor/graph_executor.cc#L256

And then, I change another things,
```
data_batch.data.CopyTo(&args_map["data"]);
data_batch.label.CopyTo(&args_map["data_label"]);
```
To
```
args_map["data"] = data_batch.data.Copy(ctx);
args_map["data_label"] = data_batch.label.Copy(ctx);
```
I got an exception; I believe it is the same thing as the original code.

---

BTW, I have to build the mxnet with debug symbol in order to trace the code to get more detail error message.

Here is my debug result:

This function call the C interface and throw the exception which is not helpful:
https://github.com/dmlc/mxnet/blob/master/cpp-package/include/mxnet-cpp/symbol.hpp#L181

Here is the line throw the exception: Error in operator conv_conv01.....
https://github.com/dmlc/mxnet/blob/master/src/c_api/c_api_symbolic.cc#L445

After I traced into this function, I gave up.
https://github.com/dmlc/nnvm/blob/b279286304ac954098d94a2695bca599e832effb/include/nnvm/pass_functions.h#L76

**Comment by NiklasTreutner** [comment]
I'm afraid i can't help you with your question, however, since you managed to get the cpp-bindings to work under win10 and msvc15, maybe you could tell me what steps you took in order to use the bindings. i'm missing the op.h (like i wrote here #5941 ) and don't know how to get it generated with the pythonWrapper.

thanks!

**Comment by shuokay** [comment]
```cpp
auto conv1 = ConvFactory(data, 64, Shape(3, 3), Shape(1, 1), Shape(1, 1), "conv01");
```
Tensor shape is `(batch, channel, height, width)`, change to
```cpp
auto conv1 = ConvFactory(data, 1, Shape(3, 3), Shape(1, 1), Shape(1, 1), "conv01");
```
may help.

**Comment by kice** [comment]
@NiklasTreutner the op.h is generated by https://github.com/dmlc/mxnet/tree/master/cpp-package/src/OpWrapperGenerator
Please use Python 2 instead of py 3.

@shuokay I think num_filter(in this case, 64) is the output of depths the Convolution. I want to input a 41x41x1 (x*y*z / w*h*d) and get 41x41x64 output.

**Comment by NiklasTreutner** [comment]
@kice thanks! is there any documentation regarding the cpp bindings that i missed?

**Comment by kice** [comment]
@NiklasTreutner the doc is inside the op.h and you may found the python doc is useful because I think they are the same thing. For everthing else, I have no idea.

**Comment by kice** [comment]
@lx75249

**Comment by lx75249** [comment]
@kice Could you give the complete code? Thanks.

**Comment by kice** [comment]
@lx75249 
```
#include <mxnet-cpp/MxNetCpp.h>

#include <string>
#include <vector>
#include <map>

#include "fmt\format.h" // https://github.com/fmtlib/fmt

using namespace mxnet::cpp;

inline Symbol ConvFactory(Symbol data, int num_filter,
						  Shape kernel,
						  Shape stride = Shape(1, 1),
						  Shape pad = Shape(0, 0),
						  const std::string & name = "")
{
	Symbol conv_w("conv_" + name + "_w"), conv_b("conv_" + name + "_b");

	Symbol conv = Convolution("conv_" + name, data,
							  conv_w, conv_b, kernel,
							  num_filter, stride, Shape(1, 1), pad);
	return Activation("relu_" + name, conv, "relu");
}

Symbol Model()
{
	Symbol data = Symbol::Variable("data");
	Symbol data_label = Symbol::Variable("data_label");

	auto conv1 = ConvFactory(data, 64, Shape(3, 3), Shape(1, 1), Shape(1, 1), "conv01");
	auto conv2 = ConvFactory(conv1 , 1, Shape(3, 3), Shape(1, 1), Shape(1, 1), "conv2");

	return LinearRegressionOutput("loss", conv2 + data, data_label);
}

int main()
{
	int batch_size = 64;
	int max_epoch = 80;
	int patch_size = 41;
	float learning_rate = 1e-4;
	float weight_decay = 1e-4;
	auto ctx = Context::gpu();
	// auto ctx = Context::cpu();

	auto model = Model();
	std::map<std::string, NDArray> args_map;
	std::map<std::string, NDArray> aux_map;

	args_map["data"] = NDArray(Shape(batch_size, 1, patch_size, patch_size), ctx);
	args_map["data_label"] = NDArray(Shape(batch_size, 1, patch_size, patch_size), ctx);
	model.InferArgsMap(ctx, &args_map, args_map);

	// use grayscale 41x41 bmp
	auto train_iter = MXDataIter("ImageRecordIter")
		.SetParam("path_imglist", "./list_folder/train.lst")
		.SetParam("path_imgrec", "./list_folder/train.rec")
		.SetParam("data_shape", Shape(1, patch_size, patch_size))
		.SetParam("batch_size", batch_size)
		.SetParam("label_width", 1)
		.CreateDataIter();

	auto val_iter = MXDataIter("ImageRecordIter")
		.SetParam("path_imglist", "./list_folder/val.lst")
		.SetParam("path_imgrec", "./list_folder/val.rec")
		.SetParam("data_shape", Shape(1, patch_size, patch_size))
		.SetParam("batch_size", batch_size)
		.SetParam("label_width", 1)
		.CreateDataIter();

	// auto *exec = model.SimpleBind(ctx, args_map);
	Optimizer* opt = OptimizerRegistry::Find("sgd");
	opt->SetParam("momentum", 0.9)
		->SetParam("rescale_grad", 1.0 / batch_size)
		->SetParam("clip_gradient", 10);

	for (int iter = 0; iter < max_epoch; ++iter) {
		LG << "Epoch: " << iter;

		train_iter.Reset();
		while (train_iter.Next()) {
			auto data_batch = train_iter.GetDataBatch();
			auto label = data_batch.label.GetShape();
			data_batch.data.CopyTo(&args_map["data"]);
			data_batch.label.CopyTo(&args_map["data_label"]);
			NDArray::WaitAll();

			auto *exec = model.SimpleBind(ctx, args_map);
			exec->Forward(true);
			exec->Backward();
			exec->UpdateAll(opt, learning_rate, weight_decay);

			delete exec;
		}

		Accuracy acu;
		val_iter.Reset();
		while (val_iter.Next()) {
			auto data_batch = val_iter.GetDataBatch();
			data_batch.data.CopyTo(&args_map["data"]);
			data_batch.label.CopyTo(&args_map["data_label"]);
			NDArray::WaitAll();

			auto *exec = model.SimpleBind(ctx, args_map);
			exec->Forward(false);
			NDArray::WaitAll();
			acu.Update(data_batch.label, exec->outputs[0]);
			delete exec;
		}
		LG << "Accuracy: " << acu.Get();
	}
	MXNotifyShutdown();
	return 0;
}
```

**Comment by lx75249** [comment]
@kice Your label size is inconsistent with the label in record io. You should genereate correct data and try again.

**Comment by kice** [comment]
```
	auto train_iter = MXDataIter("ImageRecordIter")
		.SetParam("path_imglist", "./list_folder/train.lst")
		.SetParam("path_imgrec", "./list_folder/train.rec")
		.SetParam("data_shape", Shape(1, patch_size, patch_size))
		.SetParam("batch_size", batch_size)
		.SetParam("label_width", 1)
		.CreateDataIter();

	auto train_label_iter = MXDataIter("ImageRecordIter")
		.SetParam("path_imglist", "./list_folder/train_label.lst")
		.SetParam("path_imgrec", "./list_folder/train_label.rec")
		.SetParam("data_shape", Shape(1, patch_size, patch_size))
		.SetParam("batch_size", batch_size)
		.SetParam("label_width", 1)
		.CreateDataIter();

	auto val_iter = MXDataIter("ImageRecordIter")
		.SetParam("path_imglist", "./list_folder/val.lst")
		.SetParam("path_imgrec", "./list_folder/val.rec")
		.SetParam("data_shape", Shape(1, patch_size, patch_size))
		.SetParam("batch_size", batch_size)
		.SetParam("label_width", 1)
		.CreateDataIter();

	auto val_label_iter = MXDataIter("ImageRecordIter")
		.SetParam("path_imglist", "./list_folder/val_label.lst")
		.SetParam("path_imgrec", "./list_folder/val_label.rec")
		.SetParam("data_shape", Shape(1, patch_size, patch_size))
		.SetParam("batch_size", batch_size)
		.SetParam("label_width", 1)
		.CreateDataIter();

	Optimizer* opt = OptimizerRegistry::Find("sgd");
	opt->SetParam("momentum", 0.9)
		->SetParam("rescale_grad", 1.0 / batch_size)
		->SetParam("clip_gradient", 10);

	for (int iter = 0; iter < max_epoch; ++iter) {
		LG << "Epoch: " << iter;

		train_iter.Reset();
		train_label_iter.Reset();
		while (train_iter.Next(), train_label_iter.Next()) {
			samples += batch_size;
			auto data_batch = train_iter.GetDataBatch();
			auto label_batch = train_label_iter.GetDataBatch();
			data_batch.data.CopyTo(&args_map["data"]);
			label_batch.data.CopyTo(&args_map["data_label"]);
			NDArray::WaitAll();

			auto *exec = model.SimpleBind(ctx, args_map);
			exec->Forward(true);
			exec->Backward();
			exec->UpdateAll(opt, learning_rate, weight_decay);

			auto output = exec->outputs;
			auto res = output[0].GetShape();

			delete exec;
		}
		auto toc = std::chrono::system_clock::now();

		DiffError acu;
		val_iter.Reset();
		val_label_iter.Reset();
		while (val_iter.Next(), val_label_iter.Next()) {
			auto data_batch = val_iter.GetDataBatch();
			auto label_batch = val_label_iter.GetDataBatch();
			data_batch.data.CopyTo(&args_map["data"]);
			label_batch.data.CopyTo(&args_map["data_label"]);
			NDArray::WaitAll();

			auto *exec = model.SimpleBind(ctx, args_map);
			exec->Forward(false);
			NDArray::WaitAll();
			acu.Update(label_batch.data, exec->outputs[0]);
			delete exec;
		}

		LG << "Epoch:\t" << iter << " " << "Accuracy: " << acu.Get();
	}
```
I use the new method to load a picture as a label, and it should be fine.

BTW, I found that I used first Convolution's weight twice. After fixing the line, everything works fine.

```
// model
Symbol data = Symbol::Variable("data");
Symbol data_label = Symbol::Variable("data_label");
auto conv1 = ConvFactory(data, 64, Shape(3, 3), Shape(1, 1), Shape(1, 1), "conv00"); // let's begin at 0

std::vector<Symbol> conv;
conv.push_back(conv1);
for (int i = 1; i <= 8; ++i) {
	std::string layerID = fmt::sprintf("conv%02d", i);
	Symbol convx = ConvFactory(conv.back(), 64, Shape(3, 3), Shape(1, 1), Shape(1, 1), layerID);
	conv.push_back(convx);
}
```
